plugins/QnA/models/class.qnamodel.php

```diff
@@ -0,0 +1,126 @@
+                $result = $this->recalculateDiscussionQnABatches($request['NumberOfBatchesDone'] ?? 0, $request['LatestID'] ?? 0);
+                break;
+        }
+
+        return $result;
+    }
+
+    /**
+     * Recalculate the QnA state of discussions.
+     * There is 4 possible QnA states for questions, Unanswered, Answered, Rejected and Accepted.
+     * There is 3 possible QnA states for comments, Accepted, Rejected and NULL (Untreated).
+     *
+     * @param array $discussionIDs discussions to be recalculated
+     * @throws Exception | Being thrown from the put method of the sql object
+     */
+    private function recalculateDiscussionsQnA($discussionIDs) {
+        // Updating questions with accepted answers.
+        Gdn::sql()
+            ->update('Discussion d')
+            ->join('Comment c', 'c.DiscussionID = d.DiscussionID and c.QnA = \'Accepted\'')
+            ->set('d.QnA', 'Accepted')
+            ->whereIn('d.DiscussionID', $discussionIDs)
+            ->put();
+
+        // Updating questions with no answers.
+        Gdn::sql()
+            ->update('Discussion d')
+            ->leftJoin('Comment c', 'c.DiscussionID = d.DiscussionID')
+            ->set('d.QnA', 'Unanswered')
+            ->where(['c.CommentID is null' => ''])
+            ->whereIn('d.DiscussionID', $discussionIDs)
+            ->put();
+
+        // Updating questions with untreated answers but no accepted answer.
+        Gdn::sql()
+            ->update('Discussion d')
+            ->join('Comment c', 'c.DiscussionID = d.DiscussionID and c.QnA is null')
+            ->leftJoin('Comment c1', 'c1.DiscussionID = d.DiscussionID and c1.QnA = \'Accepted\'')
+            ->set('d.QnA', 'Answered')
+            ->where(['c1.CommentID is null' => ''])
+            ->whereIn('d.DiscussionID', $discussionIDs)
+            ->put();
+
+        // Updating questions with ONLY rejected answers.
+        Gdn::sql()
+            ->update('Discussion d')
+            ->join('Comment c', 'c.DiscussionID = d.DiscussionID and c.QnA = \'Rejected\'')
+            ->leftJoin('Comment c1', 'c1.DiscussionID = d.DiscussionID and (c1.QnA = \'Accepted\' OR c1.QnA is null)')
+            ->set('d.QnA', 'Rejected')
+            ->where(['c1.CommentID is null' => ''])
+            ->whereIn('d.DiscussionID', $discussionIDs)
+            ->put();
+    }
+
+    /**
+     * Preparing batches of discussions (questions) and triggering recalculation of their QnA state.
+     * Updating records in batches of 1000 to make sure we don't lock the table for too long.
+     *
+     * @param integer $numberOfBatchesDone number of batches already processed
+     * @param integer $latestID latest discussionID we treated
+     * @return array current state of QnA recalculation. Formatted to match what "dba.js" expects
+     */
+    private function recalculateDiscussionQnABatches($numberOfBatchesDone, $latestID) {
+        $perBatch = 1000;
```

Comment:
I think this could be a constant.

Reply:
I thought about this, is there a gain to make it a constant if it's not used anywhere else ?